File: Week2/SortingProblem/SortedAndRotatedArray.java
public class SortedAndRotatedArray {
/*
 * Given an array nums, return true if the array was originally sorted in non-decreasing order, then rotated some number of positions
 * (including zero). Otherwise, return false. There may be duplicates in the original array.
 * 
 * https://leetcode.com/problems/check-if-array-is-sorted-and-rotated/
 * T(c) -> O(n), S(c) -> O(1)
 */

    static boolean checkSortedAndRotated(int[] nums) {
        int changeIndex = -1;
        // check the index where order changes in array
        for(int i=0;i<nums.length-1;i++)
        {
            if(nums[i]>nums[i+1])
            {
                changeIndex = i+1;
                break;
            }
        }
        
        // everything is sorted
        if(changeIndex == -1)
        {
            return true;
        }
        
        // check from that index everything is sorted or not (included last and first element also)
        for(int j = changeIndex; j<nums.length;j++)
        {
            if(nums[j]>nums[(j+1) % nums.length])
            {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        int arr[] = {3,4,5,1,2};
        int arr2[] = {2,1,3,4};
        System.out.println(checkSortedAndRotated(arr));
        System.out.println(checkSortedAndRotated(arr2));
    }
}

File: Week2/Python/SortingProblem/SortedAndRotatedArray.py
"""
    Given an array nums, return true if the array was originally sorted in non-decreasing order, then rotated some number 
    of positions (including zero). Otherwise, return false. There may be duplicates in the original array.

    Note: An array A rotated by x positions results in an array B of the same length such that A[i] == B[(i+x) % A.length], 
    where % is the modulo operation.

    https://leetcode.com/problems/check-if-array-is-sorted-and-rotated/
"""


def check(nums):
    """ T(c) -> O(n), S(c) -> O(1) """
    changeIdx = -1
    # find where order is changed
    for i in range(len(nums)-1):
        if nums[i] > nums[i+1]:
            changeIdx = i+1
            break
    
    # if no order changed, means array is sorted
    if changeIdx == -1:
        return True
    
    # check if order is fine from changeIndex till first element
    for i in range(changeIdx, len(nums)):
        if nums[i] > nums[(i+1) % len(nums)]:
            return False
        
    return True


arr = [3,4,5,1,2]
print(arr)
print(check(arr))
